Adoption Application

JessSumner opened the pull request:
Why:
The rescue wants all adopters to fill out an application form. They want to form to provide contact information and details on their experience with rats.

This PR:
Creates an application form using form_for helpers.
The data is saved in a adoption application model.
Contact information fields are validated as they are crucial.
Error messages will display if these fields are left empty or incorrectly filled out.
Flash messages display if the application is saved and the user is redirected to the rat_adoptions page.
If the form cannot be saved a flash message is displayed along with the errors.

JessSumner committed:
Adoption Application started

Created a model for adoption application data. Link has been made from the adopt page to the new adoption_application page. Columns have been added to the model for name, address and contact info.

JessSumner committed:
Adoption Application Form started

Why:
Want potential adopters to be able to fill out application form on the website.

This PR:
Updates the database from using sqlite to postgresql.
Adds an instance variable of adoption_application.
Uses form_for to create the form fields for name, address, city, cell and email.

JessSumner committed:
Adoption Application fields added

Why:
More questions required on the application form.

This PR:
Rolled back previous migrations and columns were added to the create table migration. This was to reduce the number of files with similar names and timestamps.
Columns were added and form fields created to obtain information the rescue needs from potential adopters.
A helper method has been created to list all the states of america so a dropdown could be provided for the state category

JessSumner committed:
Install Bitters and final lines added to form.

Why
Adoption application needs formating.

This PR.
Adds a final line which was missed in previous commit.
Bitters was installed to begin styling and formatting the application form.

JessSumner committed:
Added validations to adoption application and added a font

Why:
It is important that a valid email address is provided for a potential adopter.

This commit:
Adds a regex for a valid email and this is included in the validation in the model.
Begins to add some styling for the adoption form. The font for the header of the form and the nav bar have been customized.

JessSumner committed:
Included gem for dependent fields

Why:
It would be nice if certain form fields would be dependent on the answer to a previous question.

This commit:
Adds a the dependent-fields-rails gem and some javascript to enable this.

JessSumner committed:
Dependent fields added

Why:
Certain fields are only necessary dependent on previous questions

This commit:
Adds the questions on which other questions are dependent. It adds underscore_rails gem on which dependent_fields_rails is dependent. Removed code using foundation as the app is no longer using it.
Began formatting the application form.

JessSumner committed:
Styling for the application form

This commit:
Adds some formatting to the form.

JessSumner committed:
Adoption application can be submitted

Why:
Applications need to be saved.

This commit:
Completes the styling of the form. Two questions now have a text area which is larger than the text fields for the other questions.
Create method has been defined. If the application is complete it can be saved and the user is redirected to the adopt page. Otherwise they are returned to their form.
Flash messages set up.

JessSumner committed:
Fields with errors highlighted

Why:
Easily notify the adopter what question is missing on their application

This commit:
Adds a red border to fields with errors.
Flash message css moved into its own file from the custom file. This aids clarity for future changes.

JessSumner committed:
Error messages display on form

Why: It is helpful if the user is directed to the fields which are preventing their form from being submitted.

This commit:
Adds error messages to the fields which are validated.
The error messages display at the top of the form when the user is redirected back to the form.

MattMSumner commented on app/assets/stylesheets/base/_base.scss:
```diff
@@ -0,0 +1,15 @@
+
+@import "variables";
+
+// Neat Settings -- uncomment if using Neat -- must be imported before Neat
```
Since we've uncommented the line below, how would you feel about removing this comment?

JessSumner commented:
@MattMSumner @ kenyonj If you have a moment I would appreciate any feedback. Thank you very much.

JessSumner commented:
@kenyonj If you have a moment I would appreciate any feedback. Thank you very much.

MattMSumner commented on app/models/adoption_application.rb:
```diff
@@ -0,0 +1,18 @@
+  validates :name, presence: true
+  validates :address, presence: true
+  validates :city, presence: true
+  validates :cell, presence: true,
```
how about:

``` rb
validates :cell, presence: true, format: {
  with: VALID_PHONE_REGEX,
  message: "Please enter phone number without spaces or hyphens."
}
```

MattMSumner commented on app/views/adoption_applications/new.html.erb:
```diff
@@ -0,0 +1,120 @@
+
+<% if @adoption_application.errors.any? %>
+  <div id="error_explanation">
+    <h2><%= pluralize(@adoption_application.errors.count, "error") %> prevented your application from being submitted:</h2>
```
How would you feel about splitting this line up? i.e.

``` html.erb
<h2>
  <%= pluralize(@adoption_application.errors.count, "error") %>
  prevented your application from being submitted:
</h2>
```

MattMSumner commented on app/views/adoption_applications/new.html.erb:
```diff
@@ -0,0 +1,120 @@
+<div class="form_questions">
+  <%= form_for @adoption_application do |f| %>
+    <%= f.label "name:" %>
+    <%= f.text_field :name %><br />
```
How would you feel about removing the `<br />`s and instead sort out the spacing with css?

kenyonj replied:
You could probably also get away with just wrapping this text field in a `p` tag

MattMSumner commented on app/views/adoption_applications/new.html.erb:
```diff
@@ -0,0 +1,120 @@
+
+<div class="form_questions">
+  <%= form_for @adoption_application do |f| %>
+    <%= f.label "name:" %>
```
You'll want the label to match you text field, so this and all the labels should be the same symbol as the line below i.e.

``` html.erb
<%= f.label :name %>
```

If you don't want to have the same label as the column name you can change it using a translation: http://apidock.com/rails/ActionView/Helpers/FormHelper/label

If you want to just add a colon I think you can do that with css.

MattMSumner commented on test/fixtures/adoption_applications.yml:
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
How would you feel about removing this file entirely?

MattMSumner commented on test/models/adoption_application_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class AdoptionApplicationTest < ActiveSupport::TestCase
```
Since you haven't written any tests here, how would you feel about removing this file?

kenyonj commented on app/views/adoption_applications/new.html.erb:
```diff
@@ -0,0 +1,120 @@
+    <ul>
+    <% @adoption_application.errors.full_messages.each do |message| %>
+      <li><%= message %></li>
+    <% end %>
```
Thoughts on moving this into an errors partial? It might be useful so that you can separate it out from the `new` template.

kenyonj commented on app/views/application/_flash.html.erb:
```diff
@@ -0,0 +1,5 @@
+  <div class="flash-<%= key %>">
+    <span><%= value %></span>
+  </div>
+<% end %>
```
This is great. You can use this same idea for errors, just need to pass in the object that you will be reading the messages from.

kenyonj commented on db/migrate/20151212182619_add_columns_to_adoption_application.rb:
```diff
@@ -0,0 +1,6 @@
+  def change
+    add_column :adoption_applications, :current_rats, :string
+    add_column :adoption_applications, :previous_rats, :string
+  end
```
Since this is still a PR, what are your thoughts on modifying the original migration to include these?

kenyonj commented:
Overall this looks great. I had some minor comments. I think it might be worth looking into [i18n](http://guides.rubyonrails.org/i18n.html). It would significantly clean up your templates and make changes to your field names in that form very easy to manage. It's a little bit more advanced, so I might even recommend that you do it in a separate PR for just `i18n` integration, but it's definitely worth looking into.

JessSumner committed:
PR changes for adoption application

This commit:
Rolls back the two migrations to create one migration for the application.
Deletes unused files: fixtures and test for adoption application.
Corrects some syntaxing suggested by Hound, replacing ' with ".
Pulls out the erb for displaying errors into its own partial. This cleans up the adoption_application_new page.

JessSumner committed:
Form labels corrected and extracted into a YAML file.

Why:
The form labels are brief descriptions but do not suffice as a question for the applicant.

This commit:
The form labels were changed to match the symbol of the form field as this is the correct form.
Using a label helper the questions which are desired were extracted into a YAML file. The YAML is called form_labels.yml
A few syntax corrections made as suggested by Hound.

MattMSumner commented on app/helpers/adoption_applications_helper.rb:
```diff
@@ -0,0 +1,58 @@
+      ["Ohio", "OH"],
+      ["Oklahoma", "OK"],
+      ["Oregon", "OR"],
+      ["Pennsylvania", "PA"],
```
I think these are fine as is but if you'd like to follow Hound's advice this would be:

``` rb
%w(Pennsylvania PA)
```

JessSumner committed:
Corrections following migration merge and tweeks to css for adoption application

Why:
It is important that certain fields are validated so that an empty form cannot be submitted.

This commit:
Adds validations to the model which was empty due to the rollback and new generation of the model in a previous commit to merge the two migrations.
Css tweeks to add more padding around the radio buttons and labels.
Color for fonts on the error messages matches the flash message color.
Added a "please select" for the state, this is the new default rather than Alabama.

JessSumner committed:
Syntax changes to controller and adoption_application.css files

This commit:
Removes trailing whitespace.
Shortens lines which were over 80 characters
Amends a class name to include hyphens not underscores

JessSumner committed:
Hound corrections

Replaced underscores with hyphens
Added ; for javascript file
Removed whitespace
Ordered properties alphabetically

JessSumner committed:
Reverted some changes

Put underscores back in to fields_with_errors as that is how the class is defined.
Added message back in for to the cell validation.
Removed the variable red as no longer using it.

JessSumner committed:
Hound corrections

Removed trailing whitespace
Corrected indentation

JessSumner commented:
@kenyonj Thank you for your feedback. I will definitely investigate i18n and I will do a separate PR for it.
